File: frameworks_demo/Gym/lua/replay_buffer.lua
require 'nn'
local class = require 'class'

ReplayBuffer = class('ReplayBuffer')

function ReplayBuffer:__init(maxSize)
  self.maxSize = maxSize
  self.currentIndex = 1
  self.replayBuffer = {}
  self.observations = torch.Tensor()
  self.actions = torch.LongTensor()
  self.rewards = torch.Tensor()
  self.next_observations = torch.Tensor()
  self.mask = torch.Tensor()
end

function ReplayBuffer:insert(observation, action, reward, done, next_observation)
  self.replayBuffer[self.currentIndex] = {
    observation = observation:clone(),
    action = action,
    next_observation = next_observation:clone(),
    reward = reward,
    done = done
  }

  self.currentIndex = self.currentIndex + 1
  if self.currentIndex > self.maxSize then
    self.currentIndex = 1
  end
end

function ReplayBuffer:sample(nSamples)
  local observationSize = self.replayBuffer[1].observation:size(1)
  self.observations:resize(nSamples, observationSize)
  self.actions:resize(nSamples, 1)
  self.rewards:resize(nSamples, 1)
  self.next_observations:resize(nSamples, observationSize)
  self.mask:resize(nSamples, 1)

  for i=1,nSamples do
    local index = torch.random(#self.replayBuffer)
    self.observations[i] = self.replayBuffer[index].observation
    self.actions[i] = self.replayBuffer[index].action
    self.rewards[i] = self.replayBuffer[index].reward
    self.next_observations[i] = self.replayBuffer[index].next_observation
    self.mask[i] = self.replayBuffer[index].done and 0 or 1
  end

  return self.observations, self.actions, self.rewards, self.next_observations, self.mask
end

File: frameworks_demo/Gym/python/replay_buffer.py
from collections import deque
import numpy as np


class replay_buffer(object):

    def __init__(self, max_size):
        self.max_size = max_size
        self.deque = deque()

        self.obserbation_batch = np.array([])
        self.action_batch = np.array([])
        self.reward_batch = np.array([])
        self.next_obserbation_batch = np.array([])
        self.mask_batch = np.array([])

    def insert(self, obs, action, reward, newobs, mask):
        self.deque.append([obs, action, reward, newobs, mask])
        if len(self.deque) > self.max_size:
            self.deque.popleft()

    def sample(self, batch_size):
        indices = np.random.choice(len(self.deque), batch_size)

        self.obserbation_batch.resize(batch_size, self.deque[0][0].shape[0])
        self.action_batch.resize(batch_size, 1)
        self.reward_batch.resize(batch_size, 1)
        self.next_obserbation_batch.resize(
            batch_size, self.deque[0][0].shape[0])
        self.mask_batch.resize(batch_size, 1)

        for b in range(batch_size):
            other_b = indices[b]
            self.obserbation_batch[b] = self.deque[other_b][0]
            self.action_batch[b] = self.deque[other_b][1]
            self.reward_batch[b] = self.deque[other_b][2]
            self.next_obserbation_batch[b] = self.deque[other_b][3]
            self.mask_batch[b] = self.deque[other_b][4]

        return self.obserbation_batch, self.action_batch, self.reward_batch, self.next_obserbation_batch, self.mask_batch
